flink-ml-lib/src/main/java/org/apache/flink/ml/feature/onehotencoder/OneHotEncoder.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * An Estimator which implements the one-hot encoding algorithm.
+ *
+ * <p>See https://en.wikipedia.org/wiki/One-hot.
+ */
+public class OneHotEncoder
+        implements Estimator<OneHotEncoder, OneHotEncoderModel>,
+                OneHotEncoderParams<OneHotEncoder> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    public OneHotEncoder() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public OneHotEncoderModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        Preconditions.checkArgument(getHandleInvalid().equals(HasHandleInvalid.ERROR_INVALID));
+
+        final String[] inputCols = getInputCols();
+
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        DataStream<Tuple2<Integer, Integer>> modelData =
+                tEnv.toDataStream(inputs[0])
+                        .flatMap(new ExtractInputColsValueFunction(inputCols))
+                        .keyBy(x -> x.f0)
+                        .transform(
+                                "findMaxIndex",
+                                Types.TUPLE(Types.INT, Types.INT),
+                                new MapPartitionFunctionWrapper<>(new FindMaxIndexFunction()));
+
+        OneHotEncoderModel model =
+                new OneHotEncoderModel()
+                        .setModelData(OneHotEncoderModelData.getModelDataTable(modelData));
+        ReadWriteUtils.updateExistingParams(model, paramMap);
+        return model;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static OneHotEncoder load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    /**
+     * Extract values of input columns of input data.
+     *
+     * <p>Input: rows of input data containing designated input columns
+     *
+     * <p>Output: Pairs of column index and value stored in those columns
+     */
+    private static class ExtractInputColsValueFunction
+            implements FlatMapFunction<Row, Tuple2<Integer, Integer>> {
+        private final String[] inputCols;
+
+        private ExtractInputColsValueFunction(String[] inputCols) {
+            this.inputCols = inputCols;
+        }
+
+        @Override
+        public void flatMap(Row row, Collector<Tuple2<Integer, Integer>> collector) {
+            for (int i = 0; i < inputCols.length; i++) {
+                Number number = (Number) row.getField(inputCols[i]);
```

Comment:
Do we have any plan to support non-integer features? The common use case of one-hot encoder is for categorical features, or simple string features.

Reply:
The current implementation aligns with Spark, in which the One Hot Encoder also only supports indexed integer values. In order to support other categorical types like string, we can add another operator like StringIndexer in front of the one hot encoder.